### ISSUE TITLE
Memory unsafety in `zerocopy`

### ISSUE THREAD

**Issue by HeroicKatora** [image-rs/canvas#6]
@joshlf `zerocopy = 0.2.2` contains an implementation flaw that exposes potential undefined behaviour. One method is missing a bound required for upholding safety invariants. We don't make use of it, so we should still be good, however this implies also refraining from using it in an unsafe way in the future. This bug report reported here for lack of a bug tracker dedicated to `zerocopy`:

`LayoutVerified<B, T>::into_ref` is missing a bound of `T: FromBytes`
and this is also missing from the documentation and implementation of
its internal `deref_helper`. Otherwise, the following works and may
gravely violate the internal variants of `T`.

```
fn any_type<T>() -> &'static T {
    let storage = Box::leak(
        vec![0u8; std::mem::size_of::<T>()]
            .into_boxed_slice());
    let verified = LayoutVerified::<_, T>::new(&storage[..])
        .unwrap();
    verified.into_ref()
}
```

**Comment by joshlf** [comment]
Wow thank you so much for catching that! I've implemented a fix [here](https://fuchsia-review.googlesource.com/c/fuchsia/+/288011). Please let me know if you think I'm missing anything, or if it fully addresses the issue you've identified.

**Comment by HeroicKatora** [comment]
Looks good to me. It doesn't seem to have spread to other places as the `Deref` impl has the correct bounds. 

A note on testability: I think [miri](https://github.com/rust-lang/miri/) is already capable of interpreting the subset of Rust used in zerocopy (I may be mistaken). That at least makes it possible to verify some unsafety guarantees and it currently sometimes finds subtle aliasing issues in the standard library. Executing and verifying some tests with it wouldn't hurt. In a much more ambitious thought maybe a fuzzer could generate adverserial but safe code trying to violate one of the checks that Miri provides. But that sounds like an ambitious separate project.

**Comment by joshlf** [comment]
OK, I've published 0.2.3 and yanked 0.2.2; you should be able to close this issue. Thanks again for the report!

**Comment by HeroicKatora** [comment]
Very good, I'll close the issue together with publishing a version that has the appropriate minimum requirement.
